harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -657,6 +657,61 @@ def test_amp(self, trial_class) -> None:
         )
         controller.run()
 
+    @pytest.mark.skipif(not HAVE_APEX, reason="Apex not available")
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="no gpu available")
+    @pytest.mark.gpu
+    def test_apex_checkpointing(self, tmp_path: pathlib.Path) -> None:
+        # There will be num_checkpoints-1 assertions made at the end of this test
+        num_checkpoints = 5
+        training_metrics = {}
+        validation_metrics = {}
+        checkpoint_dir = str(tmp_path.joinpath("apex_checkpoint"))
+        latest_checkpoint = None
+        steps_completed = 0
+
+        def make_workloads(checkpoint_idx: int = 0) -> workload.Stream:
+            # Train as usual
+            trainer = utils.TrainAndValidate()
+            # In order to expect the loss to decrease after loading the first checkpoint,
+            #  make sure the scaler has had enough steps to stop reducing so that training
+            #  can actually proceed.
```

Comment:
how about just simplify the test so it doesn't make the "monotonic decreasing" assertion?  Because this is a unit test, it is reasonable that it only focuses on exactly one detail (preservation of scaler state, for instance).

The other test you wrote provides very good coverage of the loss behavior, but no coverage of scaler preservation around checkpoints.  This test could very reasonably provide the scaler preservation coverage without having to think about loss behavior.

Reply:
Because the scaler state changes during training, so reloading the same state over and over doesn't really prove anything